### PR TITLE
List view: Modify the shortcut to focus while open

### LINKED ISSUE #29466: Persistent List View: Add a way to focus back to the sidebar via keyboard
_This issue is about the Persistent List View experiment (#28637), currently only available in the Site Editor._

The List View can be toggled with a <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> shortcut (<kbd>Shift</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> on Win).
The list is navigable with the arrow keys, and selecting an item will move the focus to the corresponding block.

There is no way to move the focus back to the List View, though (unless tabbing all the way back to it).
The workaround is to toggle the sidebar off and on again, but it's obviously not ideal.

### PR DESCRIPTION
**Pull request by alexstine**
## What?

Fixes #29466

This PR does the following.
1. Changes the shortcut to focus the list view when the list view is open and focus is not inside the sidebar.
2. When the list view is open and focus is inside the sidebar, using the shortcut will close it as normal.

## Why?

See the related issue for more information. This allows the list view to remain visible for the sighted and still allows keyboard users to navigate back to the list view without closing and re-opening.

## How?

It is not a pretty solution but it works. I left comments in the code to guide. The rest is magic.

## Testing Instructions

1. This is only in the post editor for now so open the post editor.
2. Use the list view shortcut, on Windows, it is alt+shift+o.
3. Notice how the list view sidebar opens and focus is placed inside.
4. Press the shortcut again.
5. Notice how focus went back to the position from which you opened the list view.
6. Press the shortcut again and list view will open.
7. Select a block.
8. Press the shortcut again.
9. Notice how focus goes back to the currently selected block in the list view.

## Additional thoughts
- [ ] Bug: The `useFocusReturn` hook needs some refactoring or different usage for this. Since it attaches when the component mounts, it will not return focus to where the user expects if they have navigated around the list view for a while. I need some way to call the hook dynamically but my research is turning up nothing. Suggestions on how to fix this would be appreciated.

It seems it would be better to give the sidebars it's own keyboard shortcuts. That would make abstraction of future functionality easier as well. For now, I think I'll leave it be. Being able to pass things from parent to child would likely help but we're not setup for it now.

### CODE REVIEW

File: packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js [hunk 1]
```diff
@@ -35,12 +39,30 @@ export default function ListViewSidebar() {
...
 	const labelId = `edit-post-editor__list-view-panel-label-${ instanceId }`;
 
+	// This ref helps us focus the list view.
+	const listViewRef = useRef();
+	// This tracks if the list view has focus and when focus loss happens.
+	const [ listViewHasFocus, setListViewHasFocus ] = useState( false );
+	// This only fires when the list view is open because of the conditional rendering. It is the same shortcut to open but that is defined as a global shortcut and only fires when the list view is closed.
+	useShortcut( 'core/edit-post/toggle-list-view', () => {
+		// If the list view has focus, we know it is safe to close.
+		if ( listViewHasFocus ) {
```
Comment (youknowriad) [inline packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: Instead of a new state value, can we instead rely on `listViewRef.current.contains( document.activeElement )` here ?

Comment (alexstine) [reply packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: @youknowriad That looks better? Seems to work the same.

**Comment by talldan** [comment]
It works really well in my testing. It's nice how it takes advantage of the roving tab index, so you can select a block from list view, and then use the shortcut to move straight back to that block in List View.

I spotted one issue, which is that if the List view sidebar close button is focused then shortcut doesn't close the sidebar, instead it moves focus into the list view treegrid. I think that's because the `ref` is bound to an element that isn't the parent of the close button, so checking for the active element doesn't work.

Changing this `ref` would make the roving tab index thing I mentioned above not work quite so well, the close button would be the first tabbable element found when using the shortcut to open or move focus into list view. A workaround for that could be to bind two refs, one that's checked when closing list view, the other that's used to move focus into list view when opening.

It adds more complexity to the code, but it may be the best option in terms of usability.

**Comment by alexstine** [comment]
@talldan How does a solution like that look? This will need a new E2E test to make sure none of this breaks but it seems to have gotten me around the bug.

Thanks for reviewing.

**Comment by alexstine** [comment]
Okay, added a fairly complex E2E for this. It should catch mistakes but this also helped me discover one other bug.

Selecting the already selected block in the list view will not trigger focus. This is likely related to `useCallback` deps? I'll fix it in a follow-up.

Thanks.

File: packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js [hunk 2]
```diff
@@ -45,9 +45,9 @@ export default function ListViewSidebar() {
 	useShortcut( 'core/edit-post/toggle-list-view', () => {
 		// If the list view has focus, we know it is safe to close.
 		if (
-			listViewRef.current.contains(
-				listViewRef.current.ownerDocument.activeElement
-			)
+			listViewRef.current
+				.closest( '[role="region"]' )
```
Comment (youknowriad) [inline packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: Was this change needed to make sure the "close" button is within this div? 

Can we instead just move the listViewRef to the div in line 62 of this file? The goal is to avoid relying on something that is not rendered by this component itself

Comment (alexstine) [reply packages/edit-post/src/components/secondary-sidebar/list-view-sidebar.js]: @youknowriad I gave it a try and updated the E2E test. How does this look?

**Comment by youknowriad** [comment]
I'm noticing a small issue when the post is empty (no blocks at all). I load the post editor, I hit the shortcut, first it opens the list view but then it's stuck I can't close it anymore.

**Comment by alexstine** [comment]
I'll try to investigate this shortly. Going out of town for a while. Tried setting up Gutenberg local dev on my Windows machine, interested to see if maybe I can improve the experience if I happen to find bugs while I am away from my Windows and Mac working environments over the LAN. Not all of us are lucky enough to fire up Linux or have Mac, this should be interesting.

**Comment by youknowriad** [comment]
@alexstine yeah, I have trouble working when I'm not on my regular setup as well.

My codesandbox PR might help here, we'll have to see in the long run.

**Comment by alexstine** [comment]
@youknowriad I pushed a fix but that still leaves us with a problem. The `useFocusOnMount` hook does not work in a no blocks situation either because there is nothing to focus if the ListView has no tabbables.

Might be time to refactor this a bit and call useEffect to handle our focus situations in the sidebar. Hate to do it, but I need full visibility over the state. useEffect would be perfect for that since I can listen for state changes.

Thoughts?

**Comment by alexstine** [comment]
Just removed the `useFocusOnMount` hook in favor of `useEffect`. This allows me to use a callback function to determine where focus should be placed. Seems to work okay.

I need to refresh the branch later today and make sure the label changes have not broken anything.

**Comment by alexstine** [comment]
I think this is mostly good to go. It might be worth breaking up the E2E test a bit if others think that is necessary. If all looks good, I'll apply to other packages. I think I finally managed to fix all bugs.

File: packages/block-editor/src/components/list-view/index.js [hunk 3]
```diff
@@ -174,6 +174,11 @@ function ListView(
 		[ isMounted.current, draggedClientIds, expandedState, expand, collapse ]
 	);
 
+	// If there are no blocks to show, do not render the list view.
```
Comment (alexstine) [inline packages/block-editor/src/components/list-view/index.js:177]: @talldan Is there a reason this was not added previously? I found it really confusing to render an empty list view.

**Comment by alexstine** [comment]
No idea why this new test failure keeps happening. I can't see what changed that could cause it.

**Comment by alexstine** [comment]
Just tried to rebase the branch. Let's see what happens with tests.

**Comment by alexstine** [comment]
@kevin940726 Can you spot why these E2E tests might be failing?

**Comment by kevin940726** [comment]
The tests failed because JavaScript threw an error "TypeError: Cannot read properties of undefined (reading 'querySelectorAll')" which broke the editor and the test runner couldn't complete the rest of the steps. I'm not sure what that error is about though.

**Comment by alexstine** [comment]
@kevin940726 Looks like I fixed that issue but tests still unhappy. Can you spot the reason something with the selection is going wrong?

Thanks.

**Comment by kevin940726** [comment]
During the `Multi-block selection › should multi-select in the ListView component with shift + up and down keys` test, on line 819, we expect `getSelectedFlatIndices()` (whatever that means) to equal `[2, 3]`, but it just returns `1`. I'm not sure what it means though, but it seems related to this PR.

**Comment by alexstine** [comment]
@talldan or @andrewserong Any ideas as to why the above test is failing?

**Comment by alexstine** [comment]
Would be nice to get a review on this. Need some assistance debugging the test, it is more than I know how to do at this point.

CC @annezazu

**Comment by andrewserong** [comment]
Thanks for the ping, apologies for the late reply, I must have missed the ping while catching up after AFK.

> Any ideas as to why the above test is failing?

I walked through the failing test manually in the post editor and I think it's failing because of a change in this PR when the list view is opened. With this PR applied, it seems that the focus is always over the first block in the list view when it opens, rather than the block that is currently selected within the editor canvas.

For example, if I create a post containing four blocks, with the fourth block selected, if I open the list view, then on `trunk` the fourth block will be focused. However with this PR applied, it looks like focus is positioned at the first block in the list view.

Was that part of the changes in this PR intentional?

**Comment by alexstine** [comment]
@andrewserong That is not supposed to happen. Maybe I am not sending focus through the component properly? Can you spare a few moments to help me make this stable? So close.

**Comment by andrewserong** [comment]
I took another look through the code, and I'm wondering if the reason the focus is going elsewhere is because the ref is now attached to the parent document overview container, which has tabbable elements within it, whereas previously the ref was attached to the container for the list view. So on trunk, I think that meant that `useFocusOnMount` would find the first tabbable node within the list view, which I think means the currently selected block?

However now that the ref is a level higher, `listViewFocusLocation.focus();` seems to set focus to the list view itself, rather than the first tabbable node within it? I'm wondering if we need to do another lookup within `listViewFocusLocation`, so possibly something like `focus.tabbable.find( listViewFocusLocation )` and then grab the first element of that, and focus to it?

I'm a little stretched for time this week (I'm off for a public holiday tomorrow), but I should have a little time on Friday to take another look.

**Comment by alexstine** [comment]
@andrewserong I could not reproduce the same. If I select the 2nd block in the list and open the list view, I land on the 2nd block in the list view. I added `preventScroll: true` option since `useFocusOnMount` uses it.

**Comment by andrewserong** [comment]
@alexstine I'm a bit stuck attempting to re-review this one, as in testing `listViewFocusLocation` always seems to be the first item in the list for me, rather than the currently selected block. It could be something to do with the environment I'm running, but it does seem consistent with the failing e2e test.

One thing that stands out to me after another dive into the code, is that over in the `ListViewBlock` component, when it's initially mounted it calls `cellRef.current.focus` on [this line](https://github.com/WordPress/gutenberg/blob/b9821b0217d0c3d9683cbe323544a23a793f9eca/packages/block-editor/src/components/list-view/block.js#L155). As far as I can tell, prior to this PR, that's the behaviour that ensures the selected item is selected when the list view is initially opened. However, the logic in this PR searches for the first tabbable element within the list view container, which defaults to the first link, because that link has `tabIndex` set to `0`. At least, from inspecting in dev tools, that seems to be what's going on, but I very well could be wrong!

My hunch at the moment is that the focus behaviour in this PR means that the cell ref's focus isn't being called, so the focus never gets placed to the selected block in the list view, and so it doesn't receive an updated `tabIndex` value.

What does this all mean for next steps?

I'm wondering if it would help if we looked into limiting the scope of what this PR deals with so that `handleSidebarFocus` is only responsible for handling focus changes when the list view is open and not focused. That way, we could continue to rely on the behaviour in `trunk` for when the list view is initially opened, which would preserve the feature where the list view is scrolled to the currently selected block.

Now that we've passed feature freeze for 6.2, I should have a little more time to dig in and play around with some of these ideas. If you're happy for me to, I can experiment in this PR. Otherwise, I'm also happy to open up a separate PR to try that out, as I don't want to overwrite any of your good work here!

Thanks again for your patience with all this, it'll be a really good enhancement to land.

**Comment by alexstine** [comment]
@andrewserong Can you remind me which browser E2E runs in? I think it is Chrome? I generally develop in Firefox so that could be it. Let me record a video tomorrow of what my experience is and maybe you can spot something that I just do not get visually.

I think the reason this PR fell a little out of scope is because having duplicate focus everywhere just did not make much sense to me. I wanted to try to make it clear to anyone in the future exactly what managed focus when as allowing the `useFocusOnMount` hook to stay around would have caused confusion.

At least that is my take...

This whole solution really is not pretty from a code perspective.

Thanks.

**Comment by andrewserong** [comment]
> Can you remind me which browser E2E runs in?

The failing e2e test uses Puppeteer to run the tests, and Puppeteer is based on Chromium browser, so that very well could be a factor here.

Also, I know what you mean about dealing with the managed focus, it's pretty tricky for folks to follow along when logic is spread across components, or is happening in different places. It took me quite a while to spot the `cellRef` focus behaviour that happens inside the list view component, too. Hopefully if we can get everything working correctly and the tests passing, the path to tidying up the code might become clearer.

I'll continue to have a play around with this also, and I'll let you know what I find!

**Comment by andrewserong** [comment]
Update: it's pretty messy, but I've had a go at hacking on a change based on this PR over in #47901, specifically within [list-view-sidebar.js](https://github.com/WordPress/gutenberg/pull/47901/files#diff-f8cfc8062b8fadfbe2f067950e866290ab601e82aa7fead74090c4bc009dbc16). It's very much a throwaway PR, so feel free to disregard it, but for the moment I've been playing with the following:

* Use `refs` for the list view button and the outline button, so that we can focus to them explicitly rather than doing a lookup for tabbable elements.
* Use a selector query (borrowing from internal behaviour in the list view component) to find the anchor tag for the selected block if a block is selected, and then focus to it.

In manual testing, that appears to resolve the issue I ran into, however I think the tests are still failing in a different way, so might not be a reliable solution. I mostly put up the code change just in case it sparks any ideas!

I have limited time tomorrow to look into this further, so if not tomorrow, then I can continue digging in on Monday.

**Comment by alexstine** [comment]
@andrewserong I actually like the idea of doing less lookups so I removed that aspect from my PR. I still require a couple lookups but I have used more refs in my latest commit. I also made a video showing the difference between my PR and trunk. I think we'll likely have to update the E2E tests. I think I get what you are saying now.

Thoughts?

Thanks.

https://user-images.githubusercontent.com/13755480/217904144-ec3609ff-0c11-402a-84e0-8917cbf49044.mp4

**Review by andrewserong — APPROVED** [review]
Thanks so much for adding the video @alexstine, it's super helpful hearing you talk through the behaviour of the feature, and it's also nice to hear your voice! You make a good point comparing how the feature works on trunk, and honing in on the changes of this PR.

After giving this PR a re-test, I think your latest refactor has fixed up the main difference between this branch and trunk, and the previously failing e2e test is now passing. While testing manually, it seems that this branch now has the same behaviour as `trunk` for me when initially opening the list view, either by clicking on the list view button or using the keyboard shortcut. So, I think the addition of the extra refs and calling `find` on `listViewApplicationFocus` seems to have fixed up any disparity there.

I think this PR is in a good place to land now. The feature as implemented is working correctly, the e2es are passing, and otherwise the existing behaviour when opening the list view appears to be the same between trunk and this PR. I imagine some of the inconsistency on `trunk` could be to do with when the cell ref within the list view sets focus. It seems like it might be occurring inconsistently over there, but that's an issue that could be investigated separately.

Great work again, this LGTM! ✨

**Review by tellthemachines — APPROVED** [review]
Thanks for persisting with this! It's working well for me on macOS with Chrome and Safari, although it doesn't work with VoiceOver, due to the keyboard shortcut on mac being ctrl+opt+o. That is an existing issue tracked by #11154, so definitely not a blocker for this PR. 

I also checked that with the list view open, it's still possible to reach it from inside a block by Shift+Tabbing a couple of times (first Shift+Tab lands us in the block toolbar, second in the list view).

One thing that would be nice to look at in a follow-up: when moving focus from the canvas to the already open list view, as far as I can tell focus is always transferred to the block that was last focused in the list view. So, if I have a Paragraph in the list view, move focus to the canvas, and then press Enter to create a second Paragraph, if I then move focus back to the list view it lands on the first Paragraph. This is also existing behaviour: I can reproduce it on trunk when Shift+Tabbing into the open list view from the canvas.

**Comment by alexstine** [comment]
@tellthemachines Good call out. I think we'll need to figure out how to change that in the `ListView` component itself. Right now, I am just placing focus on the block with the proper tabindex in the grid, magic.

I am ready to give this a try and see if anyone notices any issues. Over the next few days, I will try to get this going for the site editor.

Thanks all.
